src/QMCDrivers/DMC/DMCUpdatePbyPL2.cpp

```diff
@@ -0,0 +1,235 @@
+}
+
+/// destructor
+DMCUpdatePbyPL2::~DMCUpdatePbyPL2() {}
+
+void DMCUpdatePbyPL2::advanceWalker(Walker_t& thisWalker, bool recompute)
+{
+  myTimers[DMC_buffer]->start();
+  Walker_t::WFBuffer_t& w_buffer(thisWalker.DataSet);
+  W.loadWalker(thisWalker, true);
+  Psi.copyFromBuffer(W, w_buffer);
+  myTimers[DMC_buffer]->stop();
+  //create a 3N-Dimensional Gaussian with variance=1
+  makeGaussRandomWithEngine(deltaR, RandomGen);
+  int nAcceptTemp(0);
+  int nRejectTemp(0);
+  //copy the old energy and scale factor of drift
+  EstimatorRealType eold(thisWalker.Properties(LOCALENERGY));
+  EstimatorRealType enew(eold);
+  RealType rr_proposed = 0.0;
+  RealType rr_accepted = 0.0;
+  RealType gf_acc      = 1.0;
+  mPosType K;
+  mTensorType D;
+  mTensorType Dchol;
+  PosType Ktmp;
+  TensorType Dtmp;
+  bool L2_proj = H.has_L2();
+  if(L2_proj)
+  {
+    Ktmp = 0.0;
+    Dtmp = 0.0;
+    for(int d=0;d<DIM;d++)
+      Dtmp(d,d) = 1.0;
+  }
+  myTimers[DMC_movePbyP]->start();
+  for (int ig = 0; ig < W.groups(); ++ig) //loop over species
+  {
+    RealType tauovermass = Tau * MassInvS[ig];
+    RealType oneover2tau = 0.5 / (tauovermass);
+    RealType sqrttau     = std::sqrt(tauovermass);
+    RealType rr;
+    for (int iat = W.first(ig); iat < W.last(ig); ++iat)
+    {
+      W.setActive(iat);
+      //get the displacement
+      GradType grad_iat = Psi.evalGrad(W, iat);
+      mPosType dr;
+      mPosType dr_diff = deltaR[iat];
+      if(!L2_proj) // normal projector
+      {
+        getScaledDrift(tauovermass, grad_iat, dr);
+        dr += sqrttau * dr_diff;
+        rr = tauovermass * dot(dr_diff, dr_diff);
+        rr_proposed += rr;
+        if (rr > m_r2max)
+        {
+          ++nRejectTemp;
+          continue;
+        }
+        if (!W.makeMoveAndCheck(iat, dr))
+          continue;
+      }
+      else // projector including L2 potential
+      {
+        // do a fake move (zero distance)
+        // this ensures the temporary distance data is correct 
+        // will need to remove this later, but requires reimplementation of computeL2DK
+        dr = 0.0;
+        if (!W.makeMoveAndCheck(iat, dr))
+          continue;
+
+        H.computeL2DK(W,iat,Dtmp,Ktmp);
+        D = Dtmp; // upcast for mixed precision
+        K = Ktmp;
+        getScaledDriftL2(tauovermass,grad_iat,D,K,dr);
+
+        W.rejectMove(iat);
+        rr = tauovermass * dot(dr_diff, dr_diff);
+        rr_proposed += rr;
+        if (rr > m_r2max)
+        {
+          ++nRejectTemp;
+          continue;
+        }
+
+        // move with just drift to update distance tables
+        if (!W.makeMoveAndCheck(iat, dr))
+          continue;
+
+        // compute diffusion step 
+        H.computeL2D(W,iat,Dtmp);
+        D = Dtmp; // upcast for mixed precision
+        Dchol = cholesky(D);
+        dr_diff = dot(Dchol,dr_diff);
+        dr += sqrttau * dr_diff;
+
+        // reverse the intermediate drift move
+        W.rejectMove(iat);
```

Comment:
How many of these early bail outs matter? How many just torture the branch prediction of the processor? 
Does it really make sense to check the move before its even scaled?

Reply:
Apart from the spurious `makeMoveAndCheck` at the beginning to ensure the temporary distances are populated, the bailout pattern is similar to the original DMC implementation (see code just below `!L2_proj`.  The check before scaling does not depend on the scaled drift in the original or the L2 algorithm.  I don't know the answer regarding necessity, but it should be the same for both algorithms.